Move src/tinyformat.h to src/tinyformat/tinyformat.h

Empact opened the pull request:
For consistency with other bundled dependencies. Having it in the top-level
directory makes it clear that it is a library that we're including
rather than a file that is a part of the project.

tinyformat/tinyformat.h and tinyformat/README.md are drawn from the existing
sha reference:
https://github.com/c42f/tinyformat/tree/689695cf58700e6defe3741829564cd682d5ae57

Existing modifications to tinyformat are moved to utilstrprintf.h, clearly
separating upstream and project code.

This includes changes in:
695041e4952ea40e0 - util: Update tinyformat
1b8fd35aadfad6a1e - Make tinyformat errors raise an exception instead of assert()ing
9b6d4c5cdc1ad7b12 - Move strprintf define to tinyformat.h
6e5fd003e04b81115 - Move `*Version()` functions to version.h/cpp
9eaa0afa6ec5d3dd0 - tinyformat: force USE_VARIADIC_TEMPLATES
b651270cd6bfdd6d7 - util: Throw tinyformat::format_error on formatting error

This does not include changes below, which are cosmetic or apparently unintentional:
64fb0ac016c7fd01c - Declare single-argument (non-converting) constructors "explicit"
4d9b4256d89d1f7c6 - Fix typos

Note I excluded only the tinyformat cpp from the boost lint checks out of an
abundance of caution - these files are not actually built into bitcoin so are
not at risk of pulling in unwanted dependencies, whereas the header, and other
dependencies might be.

Empact commented:
This is an alternative to #13845. This includes just the header and readme, while #13845 includes the full subtree - more automatic checking is possible with that, but at the cost of including support files.

laanwj commented:
Concept ACK

laanwj commented:
Closing this in favor of  #13845, at least I'd say that if we're going to move it anyhow  go the full yard and make it a subtree.

laanwj commented:
Reopening because of @MarcoFalke's concern, agree that #13845 pulls in a bizarrely high number (46!) of files. And the whole *intent* of `tinyformat.h` is to have a file that can be easily copied into a project.
It's not clear it is worth it.

Empact committed:
Move src/tinyformat.h to src/tinyformat/tinyformat.h

For consistency with other bundled dependencies. Having it in the top-level
directory makes it clear that it is a library that we're including
rather than a file that is a part of the project.

tinyformat/tinyformat.h and tinyformat/README.md are drawn from the existing
sha reference:
https://github.com/c42f/tinyformat/tree/689695cf58700e6defe3741829564cd682d5ae57

Existing modifications to tinyformat are moved to utilstrprintf.h, clearly
separating upstream and project code.

This includes changes in:
695041e4952ea40e0 - util: Update tinyformat
1b8fd35aadfad6a1e - Make tinyformat errors raise an exception instead of assert()ing
9b6d4c5cdc1ad7b12 - Move strprintf define to tinyformat.h
6e5fd003e04b81115 - Move `*Version()` functions to version.h/cpp
9eaa0afa6ec5d3dd0 - tinyformat: force USE_VARIADIC_TEMPLATES
b651270cd6bfdd6d7 - util: Throw tinyformat::format_error on formatting error

This does not include changes below, which are cosmetic or apparently unintentional:
64fb0ac016c7fd01c - Declare single-argument (non-converting) constructors "explicit"
4d9b4256d89d1f7c6 - Fix typos

Note I excluded only the tinyformat cpp from the boost lint checks out of an
abundance of caution - these files are not actually built into bitcoin so are
not at risk of pulling in unwanted dependencies, whereas the header, and other
dependencies might be.

MarcoFalke commented on src/tinyformat/README.md:
```diff
@@ -0,0 +1,452 @@
+# tinyformat.h
+
+## A minimal type safe printf() replacement
```
No need to add a third party readme. This will show up in all history/logs and greps.

MarcoFalke commented:
Concept ACK

Empact commented:
Dropped the readme. 👍

MarcoFalke commented on test/lint/README.md:
```diff
@@ -17,6 +17,7 @@ To use, make sure that you have fetched the upstream repository branch in which
 maintained:
 * for `src/secp256k1`: https://github.com/bitcoin-core/secp256k1.git (branch master)
 * for `src/leveldb`: https://github.com/bitcoin-core/leveldb.git (branch bitcoin-fork)
+* for `src/tinyformat`: https://github.com/c42f/tinyformat.git (branch master)
```
nit: Not a subtree

MarcoFalke commented:
utACK c5ee680fdfe806ab6ffbfca47f6cd28e4fb4e049

Empact commented:
Removed inaccurate test/lint/README mention and unnecessary test/lint/lint-includes.sh filter.

laanwj commented:
This was discussed at the 2018-08-30 IRC meeting and there was no agreement to do this at all.
```
21:20 < wumpus> one topic I'd like to discuss is where to move tinyformat in the source tree, if we're going to do
                that at all, I hate it when there's two competing PRs open for something
21:20  * jonasschnelli is lost in PRs
21:20 < wumpus> #topic tinyformat move
21:20 < wumpus> e.g.: #13846, #13845, or keep as is
21:20 < gribble> https://github.com/bitcoin/bitcoin/issues/13846 | Move src/tinyformat.h to
                 src/tinyformat/tinyformat.h by Empact . Pull Request #13846 . bitcoin/bitcoin . GitHub
21:20 < gribble> https://github.com/bitcoin/bitcoin/issues/13845 | Include tinyformat as a subtree by Empact . Pull
                 Request #13845 . bitcoin/bitcoin . GitHubAsset 1Asset 1
21:21 < wumpus> I'm ok with all three options but not with leaving those PRs open forever
21:22 < jonasschnelli> The subtree looked to me after an overkill,... I would prefer #13846 (no strong opinion)
21:22 < gribble> https://github.com/bitcoin/bitcoin/issues/13846 | Move src/tinyformat.h to
                 src/tinyformat/tinyformat.h by Empact . Pull Request #13846 . bitcoin/bitcoin .
21:22 < wumpus> I guess MarcoFalke is not here?
21:23 < wumpus> I think he has the strongest opinion about it
21:23 < gmaxwell> would we really do a subtree for a single file?
21:23 < wumpus> no.
21:24 < wumpus> I think this is pretty much unnecessary, and certainly the subtree one contains lots of changes
21:24 < gmaxwell> seems like change for the sake of change to me.
21:24 < wumpus> too much of that
21:25 < achow101> I'm in favor of keeping it as is
```
If there's no agreement how to go forward here, I'd say it's not extremely important to do, so I'm closing both PRs for now.

MarcoFalke commented:
Imo this could still be done the next time we touch tinyformat, but yeah no rush
